lib/LLVMIRCodeGen/AllocationsInfo.cpp

```diff
@@ -35,6 +35,45 @@ using llvm::cast;
 using llvm::dyn_cast;
 using llvm::isa;
 
+/// If \p PH is an output placeholder, \returns true.
+/// This is determined by checking if the PH has a user which uses the PH as an
+/// overwritten input.
+bool isOutputPlaceholder(const Placeholder *PH) {
+  for (const auto &use : PH->getUsers()) {
+    // Look through the inputs of the PH's users. If an input is overwritten
+    // check if it's the PH, if it is return true.
+    auto *user = use.getUser();
+    for (unsigned i = 0, numInputs = user->getNumInputs(); i < numInputs; i++) {
+      // If the input is not overwritten we can continue.
+      if (!user->isOverwrittenNthInput(i)) {
+        continue;
+      }
+      auto input = use.getUser()->getNthInput(i);
+      if (input.getNode() == PH) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+/// If \p PH is an input placeholder, \returns true.
+bool isInputPlaceholder(const Placeholder *PH) {
+  // Check that the PH is the input to a saveNode or is used by a non saveNode.
+  for (const auto &use : PH->getUsers()) {
+    // Check if PH is an input to a saveNode.
+    if (auto *save = dyn_cast<SaveNode>(use.getUser())) {
+      auto input = save->getInput();
+      // If the PH is not an input to the saveNode we keep looking.
+      if (input.getNode() != PH) {
+        continue;
+      }
+    }
+    return true;
+  }
+  return false;
+}
+
 void AllocationsInfo::allocateWeightVars(const IRFunction *F) {
```

Comment:
The fact that we still have `AllocationsInfo` which performs its own allocation worries me. It is a recipe for disaster. I thought we switched completely to `RuntimeBundle` and centralized the allocation logic there. But I see that our CPU backend currently invokes `allocateJITMemory` and `runtime::RuntimeBundle::create` from `LLVMBackend::compileIRWithoutConstants`! This seems very odd. 

@gcatron What was the reason for not switching completely to `RuntimeBundle` or at least implementing `AllocationsInfo` on top of `RuntimeBundle` in the CPU backend? Why does each of them has its own allocation logic?

Reply:
I filed #3011 to track this issue.